src/key-hotp-oracle.md

```diff
@@ -0,0 +1,865 @@
+
+The below figure illustates the way asynchronous operations are handled in Tock,
+using our encryption oracle capsule as an example. One further detail
+illustrated in this figure is the fact that providers of a given interface
+(e.g., `AES128`) may not always be able to perform a large user-space operation
+in a single call; this may be because of hardware-limitations, limited buffer
+allocations, or to avoid blocking the kernel for too long in
+software-implentations. In this case, a userspace-operation is broken up into
+multiple smaller operations on the underlying provider, and the next
+sub-operation is scheduled once a callback has been received:
+
+![An Illustration of Tock's Asynchronous Driver Model](imgs/encryption_oracle_capsule.svg)
+
+To allow our capsule to receive `crypt_done` callbacks, add the following trait
+implementation:
+
+```rust
+use kernel::hil::symmetric_encryption::Client;
+
+impl<'a, A: AES128<'a> + AES128Ctr> Client<'a> for EncryptionOracleDriver<'a, A> {
+    fn crypt_done(&'a self, mut source: Option<&'static mut [u8]>, destination: &'static mut [u8]) {
+	    unimplemented!()
+    }
+}
+```
+
+With this trait implemented, we can wire up the `oracle` driver instance to
+receive callbacks from the AES engine (`base_peripherals.ecb`) by uncommenting
+the following line in `boards/nordic/nrf52840dk/src/main.rs`:
+
+```diff
+- // Leave commented out for now:
+- // kernel::hil::symmetric_encryption::AES128::set_client(&base_peripherals.ecb, oracle);
++ kernel::hil::symmetric_encryption::AES128::set_client(&base_peripherals.ecb, oracle);
+```
+
+### Multiplexing Between Processes
+
+While our underlying `AES128` implementation can only handle one request at a
+time, multiple processes may wish to use this driver. Thus our capsule
+implements a queueing system: even when another process is already using our
+capsule to decrypt some ciphertext, another process can still initate such a
+request. We remember these requests through the `request_pending` flag in our
+`ProcessState` grant, and we've already implemented the logic to set this flag!
+
+Now, to actually implement our asynchronous decryption operation, it is further
+important to keep track of which process' request we are currently working on.
+We add an additional state field to our `EncryptionOracleDriver` holding an
+[`OptionalCell`](https://docs.tockos.org/kernel/utilities/cells/struct.optionalcell):
+this is a container whose stored value can be modified even if we only hold an
+immutable Rust reference to it. The _optional_ indicates that it behaves similar
+to an `Option` – it can either hold a value, or be empty.
+
+```diff
+  use kernel::utilities::cells::OptionalCell;
+
+  pub struct EncryptionOracleDriver<'a, A: AES128<'a> + AES128Ctr> {
+      aes: &'a A,
+      process_grants: Grant<ProcessState, UpcallCount<0>, AllowRoCount<0>, AllowRwCount<0>>,
++     current_process: OptionalCell<ProcessId>,
+  }
+```
+
```

Comment:
```suggestion
```diff
pub fn new(
        aes: &'a A,
        _source_buffer: &'static mut [u8],
        _dest_buffer: &'static mut [u8],
        process_grants: Grant<ProcessState, UpcallCount<0>, AllowRoCount<0>, AllowRwCount<0>>,
    ) -> Self {
        EncryptionOracleDriver {
            process_grants,
            aes,
+            current_process: OptionalCell::empty(),
        }
    }
```
```

Make sure you add the closing ``` here, github seems to use the ``` to as ending for the suggestion.

Reply:
> Make sure you add the closing ``` here, github seems to use the ``` to as ending for the suggestion.

10/10, thanks GitHub!